### PR TITLE
Harmonic sequence

### LINKED ISSUE #29: Harmonic Sequence

### PR DESCRIPTION
**Pull request by Omnia-Agabani**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [ ] The file name describes the function's behavior
- [ ] There is a module docstring in the function file
- [ ] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [ ] There is a module docstring in the tests file

### Unit Tests

- [ ] The test class has a helpful name in PascalCase
- [ ] The test class has a docstring
- Every unit test has
  - [ ] A helpful name
  - [ ] A clear docstring
  - [ ] Only one assertion
  - [ ] There is no logic in the unit test
- [ ] All tests pass
- [ ] There are tests for defensive assertions
- [ ] There are tests for boundary cases

### Function Docstring

- [ ] The function's behavior is described
- The function's arguments are described:
  - [ ] Type
  - [ ] Purpose
  - [ ] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [ ] Type
  - [ ] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [ ] Each assumption about an argument is checked with an assertion
  - [ ] Each assertion checks for _only one_ assumption about the argument
- [ ] Include 3 or more (passing!) doctests

### The Function

- [ ] The function's name describes it's behavior
- [ ] The function's name matches the file name
- [ ] The function has correct type annotations

### Do's

- [ ] Variable names help to understand the strategy
- [ ] Any comments are clear and describe the strategy
- [ ] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [ ] The function's strategy _is not_ described in the documentation
- [ ] Comments explain the _strategy_, **not** the _implementation_
- [ ] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [ ] The code passes the formatting checks
- [ ] The code passes all Ruff linting checks
- [ ] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [ ] Variables are named with snake_case
- [ ] Variable names are clear and helpful
- [ ] The code follows the strategy as simply as possible
- [ ] The implementation is as simple as possible given the strategy
- [ ] The code includes defensive assertions
- [ ] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/harmonic_sequence.py [hunk 1]
```diff
@@ -1,6 +1,16 @@
...
+"""Creates harmonic sequence, based on formula: A(n): 1/ (a +(n-1) *d)
+given a, d and n are integers. and returns
+the list containing the harmonic sequence.
+
+Created on 2024-12-30
+Author: Omnia Mustafa - Abdulgadir
+"""
+
+
+def harmonic_sequence(a: int, d: int, n: int) -> list:
```
Comment (Goutbi) [inline solutions/harmonic_sequence.py]: Consider making the return type annotation more specific using `list[float]` instead of just `list`:

def harmonic_sequence(a: int, d: int, n: int) -> list[float]:

File: solutions/harmonic_sequence.py [hunk 2]
```diff
@@ -10,4 +20,62 @@ def harmonic_sequence(a, d, n):
...
+    >>> harmonic_sequence(5, "Omnia", 10)
+    AssertionError: D must be integer
+    >>> harmonic_sequence(1, 1, [1, 2, 3])
+    AssertionError: N must be integer
     """
+    # Assert an error if any input is not integer
+    assert isinstance(a, int), "A must be integer"
+    assert isinstance(d, int), "D must be integer"
+    assert isinstance(n, int), "N must be integer"
+
```
Comment (Goutbi) [inline solutions/harmonic_sequence.py]: I suggest adding a Raises section to the function's docstring to better document the assertions. Here's how you could add it:

    Raises:
    -------
    AssertionError
        If a is not an integer
        If d is not an integer
        If n is not an integer

**Review by Goutbi — APPROVED** [review]

**Review by Goutbi — APPROVED** [review]

**Review by Goutbi — APPROVED** [review]

**Comment by Goutbi** [comment]
Well-structured code with comprehensive tests. Clean and efficient implementation overall.
